src/backend/distributed/transaction/transaction_management.c

```diff
@@ -913,3 +946,229 @@ EnsurePrepareTransactionIsAllowed(void)
+ */
+static HTAB *
+ParentTransactionPropagatedObjects(bool readonly)
+{
+	if (activeSubXactContexts == NIL)
+	{
+		return PropagatedObjectsInTx;
+	}
+
+	int nestingLevel = list_length(activeSubXactContexts);
+	if (nestingLevel == 1)
+	{
+		/* the parent is the root transaction, when there is single level sub-transaction */
+		return PropagatedObjectsInTx;
+	}
+
+	/* parent is upper sub-transaction */
+	Assert(nestingLevel >= 2);
+	SubXactContext *state = list_nth(activeSubXactContexts, nestingLevel - 2);
+	if (state->propagatedObjects == NULL && !readonly)
+	{
+		/* lazily create hashset for parent sub-transaction */
+		state->propagatedObjects = CreateSubtransactionPropagatedObjectsHash();
+	}
+	return state->propagatedObjects;
+}
+
+
+/*
+ * MovePropagatedObjectsToParentTransaction moves all objects propagated in the current
+ * sub-transaction to the parent transaction. This should only be called when there is
+ * active sub-transaction.
+ */
+static void
+MovePropagatedObjectsToParentTransaction(void)
+{
+	Assert(llast(activeSubXactContexts) != NULL);
+	HTAB *currentPropagatedObjects = CurrentTransactionPropagatedObjects(true);
+	if (currentPropagatedObjects == NULL)
+	{
+		/* nothing to move */
+		return;
+	}
+
+	/* 
+	 * Only after we know we have objects to move into the parent do we get a handle on
+	 * a guaranteed existing parent hash table. This makes sure that the parents only
+	 * get populated once there are objects to be tracked.
+	 */
+	HTAB *parentPropagatedObjects = ParentTransactionPropagatedObjects(false);
+
+	HASH_SEQ_STATUS propagatedObjectsSeq;
+	hash_seq_init(&propagatedObjectsSeq, currentPropagatedObjects);
+	ObjectAddress *objectAddress = NULL;
+	while ((objectAddress = hash_seq_search(&propagatedObjectsSeq)) != NULL)
+	{
+		hash_search(parentPropagatedObjects, objectAddress, HASH_ENTER, NULL);
+	}
+}
+
+
+/*
+ * CreateSubtransactionPropagatedObjectsHash creates a hash table used to keep
+ * track of the objects propagated in the current subtransaction.
+ */
+static HTAB *
+CreateSubtransactionPropagatedObjectsHash(void)
+{
+	HASHCTL info;
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(ObjectAddress);
+	info.entrysize = sizeof(ObjectAddress);
+	info.hash = tag_hash;
+	info.hcxt = CitusXactCallbackContext;
+
+	int hashFlags = (HASH_ELEM | HASH_CONTEXT | HASH_FUNCTION);
+	return hash_create("Subtx-%d Propagated Objects", 16,
+					   &info, hashFlags);
+}
+
+
+/*
+ * InitTransactionPropagatedObjectsHash initiates a hash table used to keep
+ * track of the objects propagated in the root transaction.
+ */
+void
+InitTransactionPropagatedObjectsHash(void)
+{
+	HASHCTL info;
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(ObjectAddress);
+	info.entrysize = sizeof(ObjectAddress);
+	info.hash = tag_hash;
+	info.hcxt = TopMemoryContext;
+
+	int hashFlags = (HASH_ELEM | HASH_CONTEXT | HASH_FUNCTION);
+	PropagatedObjectsInTx = hash_create("Tx Propagated Objects", 16,
+										&info, hashFlags);
```

Comment:
Since we allocate exactly one hash table for the length of the backend, I wonder if this could keep an excessive reservation on memory if a session is used for a transaction performing a lot of ddl commands, eg. a migration.

On its own this wouldn't cause issues, however if this is done via a pooler a very sizable table could be kept in memory quite long.

How complex do you think it would be to default the upper tx based hash table to a `NULL` pointer and only create a table when a transaction involves (starts tracking object addresses) ddl commands that we want to track>

Reply:
addressed it